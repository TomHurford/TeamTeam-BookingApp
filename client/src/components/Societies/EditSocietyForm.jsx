import React from "react";

import { useFormik } from "formik";
import * as Yup from "yup";
import EditSocietyCommittee from "./ChangeSocietyMembers/EditSocietyCommittee";
import "../../styles/Society.css";
const jwtController = require("../../utils/jwt.js");
import { useParams } from "react-router-dom";

function EditSocietyForm() {
  const formik = useFormik({
    initialValues: {
      societyId: parseInt(useParams().id),
      societyName: "",
      category: "",
      societyEmail: "",
      description: "",
      website: "",
      instagram: "",
      twitter: "",
      facebook: "",
      logo: "",
      banner: "",
    },

    validationSchema: Yup.object({
      societyId: Yup.number(),
      societyName: Yup.string(),
      societyEmail: Yup.string().email("Must be a valid email address"),
      description: Yup.string().min(
        50,
        "Society description must be at least 50 characters."
      ),
      website: Yup.string().url("Must be a valid URL"),
      instagram: Yup.string().url("Must be a valid URL"),
      twitter: Yup.string().url("Must be a valid URL"),
      facebook: Yup.string().url("Must be a valid URL"),
      logo: Yup.string().url("Must be a valid URL"),
      banner: Yup.string().url("Must be a valid URL"),
    }),

    onSubmit: (values) => {
      if (
        values.societyName === "" &&
        values.category === "" &&
        values.description === "" &&
        values.website === "" &&
        values.instagram === "" &&
        values.twitter === "" &&
        values.facebook === "" &&
        values.logo === "" &&
        values.banner === "" &&
        values.societyEmail === ""
      ) {
        alert("Please fill in at least one field");
      } else {
        const data = {
          societyId: parseInt(values.societyId),
          name: values.societyName,
          category: values.category,
          email: values.societyEmail,
          description: values.description,
          links: {
            website: values.website,
            instagram: values.instagram,
            twitter: values.twitter,
            facebook: values.facebook,
            logo: values.logo,
            banner: values.banner,
          },
        };
        console.log(jwtController.getToken());
        console.log(values);
        console.log(data);
        fetch("http://localhost:5001/societies/updateSociety", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + jwtController.getToken(),
          },
          body: JSON.stringify(data),
        })
          .then((res) => res.status)
          .then((status) => {
            if (status === 200) {
              alert("Society updated successfully!");
            } else {
              alert("Error updating society");
            }
          });
      }
    },
  });

  return (
    <div style={{ marginTop: "60px", marginLeft: "8px" }}>
      <h1>Edit Society</h1>

      <form onSubmit={formik.handleSubmit}>
        <div className="form-group">
          <label htmlFor="societyName">Society Name</label>
          <input
            name="societyName"
            value={formik.values.societyName}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
          />
          {formik.touched.societyName && formik.errors.societyName ? (
            <label className="errortext" htmlFor="societyName">
              {formik.errors.societyName}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>

        <div className="form-group">
          <label>Select a category</label>
          <select
            id="category"
            name="category"
            value={formik.values.category}
            onChange={formik.handleChange}
            className="form-select"
          >
            <option defaultValue>Other</option>
            <option>Academic</option>
            <option>Social</option>
            <option>Sports</option>
          </select>
        </div>

        {/*Email below*/}
        <div className="form-group">
          <label htmlFor="email">Society Email</label>
          <input
            name="email"
            value={formik.values.email}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
          />
          {formik.touched.societyEmail && formik.errors.societyEmail ? (
            <label className="errortext" htmlFor="societyEmail">
              {formik.errors.societyEmail}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>
        {/*Email Above */}

        {/* Socials below*/}
        <h5>Socials</h5>

        <div className="form-group">
          <label htmlFor="website">Website</label>
          <input
            name="website"
            value={formik.values.website}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
          />
          {formik.touched.website && formik.errors.website ? (
            <label className="errortext" htmlFor="website">
              {formik.errors.website}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>

        <div className="form-group">
          <label htmlFor="instagram">Instagram</label>
          <input
            name="instagram"
            value={formik.values.instagram}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
          />
          {formik.touched.instagram && formik.errors.instagram ? (
            <label className="errortext" htmlFor="instagram">
              {formik.errors.instagram}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>

        <div className="form-group">
          <label htmlFor="twitter">Twitter</label>
          <input
            name="twitter"
            value={formik.values.twitter}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
          />
          {formik.touched.twitter && formik.errors.twitter ? (
            <label className="errortext" htmlFor="twitter">
              {formik.errors.twitter}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>

        <div className="form-group">
          <label htmlFor="facebook">Facebook</label>
          <input
            name="facebook"
            value={formik.values.facebook}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
          />
          {formik.touched.facebook && formik.errors.facebook ? (
            <label className="errortext" htmlFor="facebook">
              {formik.errors.facebook}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>

        {/* Image links below*/}
        <h5>Images</h5>
        <small id="imageHelp" className="form-text text-muted">
          You must enter a link to the images
        </small>
        <div className="form-group">
          <label htmlFor="logo">Society Logo</label>
          <input
            name="logo"
            value={formik.values.logo}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
          />
          {formik.touched.logo && formik.errors.logo ? (
            <label className="errortext" htmlFor="logo">
              {formik.errors.logo}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>

        <div className="form-group">
          <label htmlFor="banner">Society Banner</label>
          <input
            name="banner"
            value={formik.values.banner}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
          />
          {formik.touched.banner && formik.errors.banner ? (
            <label className="errortext" htmlFor="banner">
              {formik.errors.banner}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>

        <div className="form-group">
          <label htmlFor="description">Society Description</label>
          <textarea
            id="description"
            name="description"
            value={formik.values.description}
            onChange={formik.handleChange}
            type="text"
            className="form-control"
            onBlur={formik.handleBlur}
            rows="8"
            cols="173"
            style={{ marginBottom: "8px" }}
          ></textarea>
          {formik.touched.description && formik.errors.description ? (
            <label className="errortext" htmlFor="description">
              {formik.errors.description}
              <br />
            </label>
          ) : (
            ""
          )}
        </div>

        <button className="btn btn-primary">Create Society</button>
      </form>

      <EditSocietyCommittee societyId={formik.values.societyId} />
    </div>
  );
}

export default EditSocietyForm;
