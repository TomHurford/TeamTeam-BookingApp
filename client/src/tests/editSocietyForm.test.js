import React from 'react';
import renderer from 'react-test-renderer';
import EditSocietyForm from '../components/Societies/EditSocietyForm';
// Snapshot test for the form to edit a society
test('EditSocietyForm snapshot test', () => {
  const component = renderer.create(<EditSocietyForm />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});