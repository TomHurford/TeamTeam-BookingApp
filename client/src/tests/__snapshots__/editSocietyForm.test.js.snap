// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSocietyForm snapshot test 1`] = `
<div
  className="page-container"
>
  <div
    className="underlay"
  />
  <div
    className="container"
  >
    <h1>
      Edit Society
    </h1>
    <form
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <label
          htmlFor="societyName"
        >
          Society Name
        </label>
        <input
          className="form-control"
          name="societyName"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="field"
      >
        <label>
          Select a category
        </label>
        <select
          className="form-select"
          id="category"
          name="category"
          onChange={[Function]}
          value=""
        >
          <option
            defaultValue={true}
          />
          <option>
            Academic
          </option>
          <option>
            Social
          </option>
          <option>
            Sports
          </option>
          <option>
            Other
          </option>
        </select>
      </div>
      <div
        className="field"
      >
        <label
          htmlFor="societyEmail"
        >
          Society Email
        </label>
        <input
          className="form-control"
          name="societyEmail"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <h5>
        Socials
      </h5>
      <div
        className="field"
      >
        <label
          htmlFor="website"
        >
          Website
        </label>
        <input
          className="form-control"
          name="website"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="field"
      >
        <label
          htmlFor="instagram"
        >
          Instagram
        </label>
        <input
          className="form-control"
          name="instagram"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="field"
      >
        <label
          htmlFor="twitter"
        >
          Twitter
        </label>
        <input
          className="form-control"
          name="twitter"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="field"
      >
        <label
          htmlFor="facebook"
        >
          Facebook
        </label>
        <input
          className="form-control"
          name="facebook"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <h5>
        Images
      </h5>
      <small
        className="form-text text-muted"
        id="imageHelp"
      >
        You must enter a link to the images
      </small>
      <div
        className="field"
      >
        <label
          htmlFor="logo"
        >
          Society Logo
        </label>
        <input
          className="form-control"
          name="logo"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="field"
      >
        <label
          htmlFor="banner"
        >
          Society Banner
        </label>
        <input
          className="form-control"
          name="banner"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div
        className="field"
      >
        <label
          htmlFor="description"
        >
          Society Description
        </label>
        <textarea
          className="form-control"
          cols="173"
          id="description"
          name="description"
          onBlur={[Function]}
          onChange={[Function]}
          rows="8"
          style={
            Object {
              "marginBottom": "8px",
            }
          }
          type="text"
          value=""
        />
      </div>
      <button
        className="button"
        type="submit"
      >
        Edit Society
      </button>
    </form>
    <hr />
  </div>
  <div>
    <div
      style={
        Object {
          "backgroundColor": "#abc4ff",
          "borderColor": "gray",
          "borderRadius": 20,
          "borderWidth": 2,
          "marginBottom": "20px",
          "marginTop": "20px",
          "width": "50%",
        }
      }
    >
      <h1
        style={
          Object {
            "marginLeft": "8px",
          }
        }
      >
        Committee Members
      </h1>
      <form
        onSubmit={[Function]}
        style={
          Object {
            "marginLeft": "8px",
          }
        }
      >
        <input
          className="addmemberinput"
          onChange={[Function]}
          placeholder="Enter member's email"
          required={true}
          style={
            Object {
              "borderRadius": 3,
            }
          }
          type="email"
          value=""
        />
        <button
          className="button"
          style={
            Object {
              "marginBottom": "8px",
              "marginLeft": "8px",
            }
          }
        >
          Add Member
        </button>
      </form>
    </div>
  </div>
</div>
`;
